Update eslint-plugin-jest to 27.9.0, switch to recommended rules only

ravicious opened the pull request:
https://github.com/jest-community/eslint-plugin-jest/blob/main/CHANGELOG.md#2600-2022-01-24
https://github.com/jest-community/eslint-plugin-jest/blob/main/CHANGELOG.md#2700-2022-08-28

I switched `plugin:jest/all` to `plugin:jest/recommended`, since /all just activates every rule in the package. This might have been fine when this package was added a couple of years ago and had a limited number of rules, but it doesn't make sense anymore.

27.0.0 requires parserOptions.project to be set, but this returned an error:

> 0:0  error  Parsing error: ESLint was configured to run on
> `<tsconfigRootDir>/web/packages/teleport/vite.config.mts` using `parserOptions.project`:
> /users/rav/src/teleport/tsconfig.json
> However, that TSConfig does not include this file. Either:
> - Change ESLint's list of included files to not include this file
> - Change that TSConfig to include this file
> - Create a new TSConfig that includes this file and include it in your parserOptions.project
> See the typescript-eslint docs for more info: https://typescript-eslint.io/linting/troubleshooting#i-get-errors-telling-me-eslint-was-configured-to-run--however-that-tsconfig-does-not--none-of-those-tsconfigs-include-this-file

I updated tsconfig.node.json so that it applies only to .mts files, as I understand that was the original intention.

ravicious committed:
Update eslint-plugin-jest to 27.9.0

https://github.com/jest-community/eslint-plugin-jest/blob/main/CHANGELOG.md#2600-2022-01-24
https://github.com/jest-community/eslint-plugin-jest/blob/main/CHANGELOG.md#2700-2022-08-28

I switched plugin:jest/all to plugin:jest/recommended, since /all just
activates every rule in the package. This might have been fine when this
package was added a couple of years ago, but it doesn't make sense anymore.

27.0.0 requires parserOptions.project to be set, but this returned an error:

    0:0  error  Parsing error: ESLint was configured to run on
    `<tsconfigRootDir>/web/packages/teleport/vite.config.mts` using `parserOptions.project`:
    /users/rav/src/teleport/tsconfig.json
    However, that TSConfig does not include this file. Either:
    - Change ESLint's list of included files to not include this file
    - Change that TSConfig to include this file
    - Create a new TSConfig that includes this file and include it in your parserOptions.project

I updated tsconfig.node.json so that it applies only to .mts files, as
I understand that was the original intention.

ravicious committed:
Autofix jest/valid-title

ravicious commented:
@ryanclark I'm requesting your review here because of the changes to tsconfig.node.json.

ryanclark commented:
> @ryanclark I'm requesting your review here because of the changes to tsconfig.node.json.

LGTM, cheers

ravicious commented:
I'm going to keep the changes related to tsconfig.node.json, however…

I made them in the first place to unblock the `unbound-method` rule of eslint-plugin-jest. [As the changelog for 27.0.0 says](https://github.com/jest-community/eslint-plugin-jest/blob/main/CHANGELOG.md#2700-2022-08-28):

> unbound-method: errors thrown by the `unbound-method` base rule are no longer suppressed - really this means that if you don't specify `project` when this rule is enabled and `@typescript-eslint/eslint-plugin` is present, that error will no longer be suppressed instead of silently doing nothing; it will still not throw if this rule is enabled without the base rule being present

However, the only reason we ran the unbound-method rule is because we used all rules from eslint-plugin-jest. It is not a recommended rule.

Setting `parserOptions.project` to `['tsconfig.json', 'tsconfig.node.json']` adds 20s to `yarn eslint`, both on my machine and CI. I'm going to remove it to keep `yarn eslint` fast.

ravicious committed:
Do not set parserOptions.project

It makes ESLint slower by ~20s and it is not needed for recommended rules
from eslint-plugin-jest.

gzdunek commented:
>Setting parserOptions.project to ['tsconfig.json', 'tsconfig.node.json'] adds 20s to yarn eslint, both on my machine and CI. I'm going to remove it to keep yarn eslint fast.

Ok, makes sense 👍
